# 4.6.2 Combining and Exporting data, part 2
##  -drop column
## -inner join 'df_prods' with 'df_orders_products-combined': using 'product_id' column
## -use _merge flag
## -Export as 'ords_prods_merge.pkl'

## (#3) Import libraries & new data set 

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#import path
path = r'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [3]:
#import cleaned products dataset 'df_prods_checked.csv'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_checked.csv'), index_col = False)

In [4]:
#import new data called, 'orders_produced_combined'
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

### (#4) Check the dimensions of imported dataframes (should be the same, always check) 

In [5]:
#check shape 
df_orders_products_combined.shape

(32434489, 11)

In [6]:
df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2539329        1             1                   2           8   
2   2539329        1             1                   2           8   
3   2539329        1             1                   2           8   
4   2539329        1             1                   2           8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN  First Order         196                  1   
1                     NaN  First Order       14084                  2   
2                     NaN  First Order       12427                  3   
3                     NaN  First Order       26088                  4   
4                     NaN  First Order       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [7]:
#check shape of products 
df_prods.shape

(49672, 6)

In [8]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [9]:
#drop unecessary columns
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

In [10]:
#check oupt now, after dropping column
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

###  (#5) combine 'df_prods' with 'df_orders_products-combined': both share 'product_id' column to use as a key and this merge includes a default 'inner join'

In [16]:
df_orders_products_merged = df_orders_products_combined.merge(df_prods, on = ['product_id'])

In [17]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order      first_order  product_id  add_to_cart_order  \
0                     NaN      First Order         196                  1   
1                    15.0  Repeat Customer         196                  1   
2                    21.0  Repeat Customer         196                  1   
3                    29.0  Repeat Customer         196                  1   
4                    28.0  Repeat Customer         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices  
0          0   both         Soda        77              7     9.0  
1          1   both         Soda        77              7     9.0  
2          1   both         Soda        77              7     9.0  
3          1   both         Soda        77              7     9.0  
4          1   both         Soda        77              7     9.0

In [18]:
#check the shape
df_orders_products_merged.shape

(32404859, 15)

### (#6) Confirm results of the merge by using a merge flag

In [19]:
df_orders_products_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### this shows that both tables have 32404859 values in the flagged _merge column

### (#7) Export the newly created dataframe as ords_prods_merge in a suitable format (taking into consideration the size, using pickle).

In [20]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))